definitions/procedures/pulp/remove.rb

```diff
@@ -0,0 +1,100 @@
+module Procedures::Pulp
+  class Remove < ForemanMaintain::Procedure
+    metadata do
+      description 'Remove pulp2'
+      for_feature :pulp2
+    end
+
+    def sys
+      ForemanMaintain::Utils::SystemHelpers.new
+    end
+
+    def pulp_data_dirs
+      [
+        feature(:pulp2).data_dir + '/published',
+        feature(:pulp2).data_dir + '/content',
+        feature(:pulp2).data_dir + '/importers',
+        feature(:pulp2).data_dir + '/uploads'
+      ]
+    end
+
+    def pulp_packages
+      [
+        'pulp-server', 'python-pulp-streamer', 'pulp-puppet-plugins',
+        'python-pulp-rpm-common', 'python-pulp-common',
+        'pulp-selinux', 'python-pulp-oid_validation',
+        'python-pulp-puppet-common', 'python-pulp-repoauth',
+        'pulp-rpm-plugins'
+      ]
+    end
+
+    def mongo_packages
+      find_installed_packages('rh-mongodb34*').concat(
+        find_installed_packages('mongodb*')
+      ).uniq
+    end
+
+    def find_installed_packages(pattern)
+      cmd = "rpm -qa --queryformat '\n%{NAME} ' | grep '#{pattern}'"
+      status, result = sys.execute_with_status(cmd)
+      if status == 0
+        result.split
+      end
+    end
+
+    def ask_to_proceed
+      question = "\nWARNING: All pulp2 packages will be removed with the following commands:" \
+        "\n# yum remove #{pulp_packages.join('  ')}" \
+        "\n# yum remove rh-mongodb34-* mongodb*" \
+        "\nAll pulp2 data will be removed. "
+      question += pulp_data_dirs.collect { |dir_name| "\n# rm -rf #{dir_name}" }.join
+      question += "\nDo you want to proceed?"
+      answer = ask_decision(question, actions_msg: 'y(yes), q(quit)')
+      abort! if answer != :yes
+    end
+
+    def run
+      ask_to_proceed
+
+      remove_pulp
+
+      remove_mongo
+
+      delete_pulp_data
+    end
+
+    def remove_pulp
+      with_spinner('Stopping pulp2 services') do |spinner|
+        pulp_services = %w[pulp_celerybeat pulp_workers pulp_resource_manager]
+        feature(:service).handle_services(spinner, 'stop', :only => pulp_services)
+      end
+
+      with_spinner('Removing pulp2 packages') do
+        packages_action(:remove, pulp_packages, :assumeyes => true)
+      end
+    end
+
+    def remove_mongo
+      with_spinner('Stopping mongo services') do |spinner|
+        mongo_services = feature(:mongo).services
+        feature(:service).handle_services(spinner, 'stop', :only => mongo_services)
+      end
+
+      with_spinner('Removing mongo packages') do
+        packages_action(:remove, mongo_packages, :assumeyes => true)
+      end
+    end
+
+    def delete_pulp_data
+      pulp_data_dir_path = feature(:pulp2).data_dir
+      with_spinner('Deleting pulp2 data directories') do |spinner|
+        pulp_data_dirs.each do |dir|
+          if File.directory?("#{pulp_data_dir_path}#{dir}")
+            execute!("rm -rf #{pulp_data_dir_path}#{dir}")
```

Comment:
Since an rm -rf is extremely dangerous, i'd like to be very careful here.  Would it be possible to generate the commands to be run as strings, print them out and then execute the same exact strings? 

This would prevent something like '# rm -rf /' from being executed if some code went really wrong, at least without the user seeing it first.

The way that the current code generates the commands as output separately and then re-generates them for execution gives me a little concern that someone could update something down the line and accidentally introduce something really bad

Reply:
That's a good point, let me try a slighter different take as per your suggestion.